### PR TITLE
Improve dot output slightly, to make ports easier to click in my tool

### PR DESCRIPTION
**Pull request by haved**
While debugging other things, I caved and put together a little web-app to render the Graphviz output, where I can click the nodes, ports and edges to look at their attributes, and highlight related elements. It is lacking features, and doesn't look good (highlight colors are extremely arbitrary), but it did help me debug :)

Note that the webapp should be able to render any Graphviz, not just output from the `GraphWriter`. The utility of the tool is of course greater when nodes and edges have non-standard attributes.

This PR gives nodes a white fill instead of being transparent, making them easier to click (can click anywhere instead of on text or border). The rest of the PR is just cleanup of the code used to dump the PointerObjectSet subset graph.

If you want to try the webapp, it's hosted at https://dot-tree-viz.vercel.app/.

The code I wrote to dump RVSDG using the GraphWriter is not included, as it was very ad hoc (see e.g. the type graph that does no de-duplication).

<details><summary>Here is some example GraphViz</summary>

```
digraph graph0 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  label="Type graph"
  
  node0 [label=ptr tooltip="555d654afb10" ];
  node1 [label=ptr tooltip="555d6547ca90" ];
  node2 [label=iostate tooltip="555d65484230" ];
  node3 [label=mem tooltip="555d65484290" ];
  node4 [label=ptr tooltip="555d6549b3a0" ];
  node5 [label=bit32 tooltip="555d6547ead0" ];
  node6 [label=bit32 tooltip="555d65476b40" ];
  node7 [label=bit32 tooltip="555d65499260" ];
  node8 [label=bit32 tooltip="555d65499040" ];
  node9 [label=vararg tooltip="555d6547cc20" ];
  node10 [label=bit32 tooltip="555d65474c20" ];
  node11 [label=ptr tooltip="555d6546f9b0" ];
  node12 [label=mem tooltip="555d6546fa80" ];
  node13 [label=mem tooltip="555d6547ca30" ];
  node14 [label=ptr tooltip="555d65487a80" ];
  node15 [label=ptr tooltip="555d6548f240" ];
  node16 [label=struct tooltip="555d65471670" ];
  node17 [label=bit32 tooltip="555d654afaa0" ];
  node18 [label=ptr tooltip="555d65483c50" ];
  node19 [label=iostate tooltip="555d65471760" ];
  node20 [label=mem tooltip="555d65471810" ];
  node21 [label=bit32 tooltip="555d654ab7b0" ];
  node22 [label=ptr tooltip="555d65480ab0" ];
  node23 [label=mem tooltip="555d65476920" ];
  node24 [label=mem tooltip="555d654773f0" ];
  node17 -> node16[id=edge0 ];
  node18 -> node16[id=edge1 ];
}
digraph graph1 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  label="RVSDG root graph"
  tooltip="555d65497890" 
  {
    rank=source;
    a0 [label=a0 type=node0 tooltip="555d65488270" ];
  }
  node25 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i0" BGCOLOR="#FFFFFF" tooltip="555d6549b2d0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">LAMBDA[reorder_basic_blocks]</TD></TR>
                        <TR><TD>
                                <TABLE BORDER="0" CELLSPACING="4" CELLPADDING="2"><TR>
                                        <TD BORDER="1" STYLE="ROUNDED" WIDTH="40" BGCOLOR="white" SUBGRAPH="graph2">graph2</TD>
                                </TR></TABLE>
                        </TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o0" type="node1" BGCOLOR="#FFFFFF" tooltip="555d65477610"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d6549b420" ];
  {
    rank=sink;
    r2 [label=r2 tooltip="555d6547e270" ];
  }
  a0 -> node25:i0:n[id=edge25 ];
  node25:o0:s -> r2[id=edge26 ];
}
digraph graph2 {
  node[shape=box style=filled fillcolor=white];
  penwidth=6;
  tooltip="555d65493e20" 
  {
    rank=source;
    a1 [label=a1 type=node2 tooltip="555d65489b00" ];
    a2 [label=a2 type=node3 tooltip="555d65474340" ];
    a1 -> a2[style=invis];
    a3 [label=a3 type=node4 tooltip="555d65488150" ];
    a2 -> a3[style=invis];
  }
  node26 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BITS32(1)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o1" type="node5" BGCOLOR="#FFFFFF" tooltip="555d6547ea40"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d65476bc0" ];
  node27 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BITS32(0)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o2" type="node6" BGCOLOR="#FFFFFF" tooltip="555d65476ab0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d654769a0" ];
  node28 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BITS32(0)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o3" type="node7" BGCOLOR="#FFFFFF" tooltip="555d654991d0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d654990c0" ];
  node29 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BITS32(0)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o4" type="node8" BGCOLOR="#FFFFFF" tooltip="555d654719d0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d654718a0" ];
  node30 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">VALIST</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o5" type="node9" BGCOLOR="#FFFFFF" tooltip="555d6547cb90"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d654897a0" ];
  node31 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">BITS32(1)</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o6" type="node10" BGCOLOR="#FFFFFF" tooltip="555d65474b90"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d65476d90" ];
  node32 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i1" BGCOLOR="#FFFFFF" tooltip="555d6546f7f0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">ALLOCA[struct]</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o7" type="node11" BGCOLOR="#FFFFFF" tooltip="555d6546f920"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o8" type="node12" BGCOLOR="#FFFFFF" tooltip="555d6546f9f0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d65476e20" ];
  node33 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i2" BGCOLOR="#FFFFFF" tooltip="555d654ae3a0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i3" BGCOLOR="#FFFFFF" tooltip="555d6547c890"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">MemStateMerge</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o9" type="node13" BGCOLOR="#FFFFFF" tooltip="555d6547c9c0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d654ae220" ];
  node34 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i4" BGCOLOR="#FFFFFF" tooltip="555d65487690"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i5" BGCOLOR="#FFFFFF" tooltip="555d654877c0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i6" BGCOLOR="#FFFFFF" tooltip="555d654878f0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">GetElementPtr</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o10" type="node14" BGCOLOR="#FFFFFF" tooltip="555d65487a10"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d654993b0" ];
  node35 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i7" BGCOLOR="#FFFFFF" tooltip="555d6547ed10"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i8" BGCOLOR="#FFFFFF" tooltip="555d6547edd0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i9" BGCOLOR="#FFFFFF" tooltip="555d6548f090"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">GetElementPtr</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o11" type="node15" BGCOLOR="#FFFFFF" tooltip="555d6548f1b0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d6547eba0" ];
  node36 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i10" BGCOLOR="#FFFFFF" tooltip="555d6547dd80"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i11" BGCOLOR="#FFFFFF" tooltip="555d6547de70"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i12" BGCOLOR="#FFFFFF" tooltip="555d6547dfa0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i13" BGCOLOR="#FFFFFF" tooltip="555d6547e0c0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">CALL</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o12" type="node16" BGCOLOR="#FFFFFF" tooltip="555d654715e0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o13" type="node19" BGCOLOR="#FFFFFF" tooltip="555d654716d0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o14" type="node20" BGCOLOR="#FFFFFF" tooltip="555d654717a0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d65473470" ];
  node37 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i14" BGCOLOR="#FFFFFF" tooltip="555d6548dfa0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">ExtractValue</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o15" type="node21" BGCOLOR="#FFFFFF" tooltip="555d654ab720"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d6548ddf0" ];
  node38 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i15" BGCOLOR="#FFFFFF" tooltip="555d6548f470"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">ExtractValue</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o16" type="node22" BGCOLOR="#FFFFFF" tooltip="555d65480a20"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d6548f2c0" ];
  node39 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i16" BGCOLOR="#FFFFFF" tooltip="555d654ab920"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i17" BGCOLOR="#FFFFFF" tooltip="555d654aba50"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i18" BGCOLOR="#FFFFFF" tooltip="555d654abb80"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">Store</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o17" type="node23" BGCOLOR="#FFFFFF" tooltip="555d654768b0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d654ab830" ];
  node40 [shape=plain style=solid label=<
<TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0">
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i19" BGCOLOR="#FFFFFF" tooltip="555d65480c70"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i20" BGCOLOR="#FFFFFF" tooltip="555d65480da0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="10"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="i21" BGCOLOR="#FFFFFF" tooltip="555d65480ed0"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="1" STYLE="ROUNDED" CELLBORDER="0" CELLSPACING="0" CELLPADDING="0" BGCOLOR="#FFFFFF" >
                        <TR><TD CELLPADDING="1">Store</TD></TR>
                </TABLE>
        </TD></TR>
        <TR><TD>
                <TABLE BORDER="0" CELLSPACING="0" CELLPADDING="0"><TR>
                        <TD WIDTH="20"></TD>
                        <TD BORDER="1" CELLPADDING="1" PORT="o18" type="node24" BGCOLOR="#FFFFFF" tooltip="555d65477380"  WIDTH="8" HEIGHT="5" FIXEDSIZE="true"></TD>
                        <TD WIDTH="20"></TD>
                </TR></TABLE>
        </TD></TR>
</TABLE>
  > tooltip="555d65480b30" ];
  {
    rank=sink;
    r0 [label=r0 tooltip="555d65477490" ];
    r1 [label=r1 tooltip="555d65477550" ];
    r0 -> r1[style=invis];
  }
  node31:o6:s -> node32:i1:n[id=edge2 ];
  node32:o8:s -> node33:i2:n[id=edge3 color="#FF0000" ];
  a2 -> node33:i3:n[id=edge4 color="#FF0000" ];
  node32:o7:s -> node34:i4:n[id=edge5 ];
  node29:o4:s -> node34:i5:n[id=edge6 ];
  node28:o3:s -> node34:i6:n[id=edge7 ];
  node32:o7:s -> node35:i7:n[id=edge8 ];
  node27:o2:s -> node35:i8:n[id=edge9 ];
  node26:o1:s -> node35:i9:n[id=edge10 ];
  a3 -> node36:i10:n[id=edge11 ];
  node30:o5:s -> node36:i11:n[id=edge12 ];
  a1 -> node36:i12:n[id=edge13 color="#00FF00" ];
  node33:o9:s -> node36:i13:n[id=edge14 color="#FF0000" ];
  node36:o12:s -> node37:i14:n[id=edge15 ];
  node36:o12:s -> node38:i15:n[id=edge16 ];
  node34:o10:s -> node39:i16:n[id=edge17 ];
  node37:o15:s -> node39:i17:n[id=edge18 ];
  node36:o14:s -> node39:i18:n[id=edge19 color="#FF0000" ];
  node35:o11:s -> node40:i19:n[id=edge20 ];
  node38:o16:s -> node40:i20:n[id=edge21 ];
  node39:o17:s -> node40:i21:n[id=edge22 color="#FF0000" ];
  node36:o13:s -> r0[id=edge23 color="#00FF00" ];
  node40:o18:s -> r1[id=edge24 color="#FF0000" ];
}
```
</details>

### CODE REVIEW

File: jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp [hunk 1]
```diff
@@ -720,7 +720,7 @@ EscapedFunctionConstraint::PropagateEscapedFunctionsDirectly(PointerObjectSet &
 }
 
 bool
-PointerObjectConstraintSet::IsFrozen()
+PointerObjectConstraintSet::IsFrozen() const
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:723]: `IsFrozen() const noexcept`

File: jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp [hunk 2]
```diff
@@ -724,7 +724,7 @@ class PointerObjectConstraintSet final
    * @return true if the constraint set has been frozen, false otherwise
    */
   [[nodiscard]] bool
-  IsFrozen();
+  IsFrozen() const;
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/PointerObjectSet.hpp:727]: `IsFrozen() const noexcept`

File: jlm/util/GraphWriter.hpp [hunk 3]
```diff
@@ -131,14 +131,28 @@ class GraphElement
...
+   * Removes the association of this GraphElement to any object in the program
+   */
+  void
+  RemoveProgramObject();
+
+  /**
+   * @return true, if this GraphElement is associated with any program object
+   */
+  [[nodiscard]] bool
+  HasProgramObject() const;
```
Comment (phate) [inline jlm/util/GraphWriter.hpp:154]: `noexcept`

File: jlm/util/GraphWriter.hpp [hunk 4]
```diff
@@ -131,14 +131,28 @@ class GraphElement
    * When using program objects as attributes, the association is used to refer to
    * the unique id of its associated graph element, instead of the object's address.
    * Within a graph, only one graph element can be associated with any given program object.
+   * @tparam T the type of the program object to associate with
    * @param object the object to associate this GraphElement with
    */
+  template<typename T>
   void
-  SetProgramObject(void * object)
+  SetProgramObject(const T * object)
```
Comment (phate) [inline jlm/util/GraphWriter.hpp:139]: `SetProgramObject(const T & object)`

File: jlm/util/GraphWriter.hpp [hunk 5]
```diff
@@ -154,14 +168,24 @@ class GraphElement
...
   void
-  SetAttributeObject(const std::string & attribute, void * object);
+  SetAttributeObject(const std::string & attribute, uintptr_t object);
+
+  /**
+   * Helper for calling SetAttributeObject with a pointer to any type
+   */
+  template<typename T>
+  void
+  SetAttributeObject(const std::string & attribute, const T * object)
```
Comment (phate) [inline jlm/util/GraphWriter.hpp:185]: `SetAttributeObject(const std::string & attribute, const T & object)`

**Review by phate — APPROVED** [review]
I am not sure whether my comments regarding references vs. pointers make sense in this context, but could we use references instead of pointers where possible?

Otherwise, I have not much to add.

**Comment by haved** [comment]
I agree using references is nicer, so it has been changed.

**Pull request merged**